### ISSUE TITLE
Built-in 464XLAT implementation

### ISSUE THREAD

**Issue by SoniEx2** [systemd/systemd#23674]
**Is your feature request related to a problem? Please describe.**

Linux-based OSes don't come with 464XLAT support out of the box, meanwhile Windows does. Mobile phones support it out of the box because mobile networks told them to. While it is unlikely anyone uses systemd-networkd in their CPE/router, it would also be nice to be able to have an entirely IPv6-only LAN. This would help prepare systemd-networkd for the eventual deprecation of RFC 1918 and global routing of the 10/8, 172.16/12 and 192.168/16 subnets. It also has some minor advantages when merging multiple LANs, as the risk of ULA conflicts is significantly lower than the risk of 1918 conflicts.

**Describe the solution you'd like**

464XLAT CLAT support, out of the box.

**Describe alternatives you've considered**

N/A

**The systemd version you checked that didn't have the feature you are asking for**

systemd 251, and also systemd-git

**Comment by yuwata** [comment]
Could you elaborate more? I am not familiar with 464XLAT.
Could you suggest a setting format to support the feature?
Also, could you provide a way to support that feature with another tools, e.g. `ip` command?

**Comment by pemensik** [comment]
A link to video of the presentation: https://ripe85.ripe.net/programme/meeting-plan/ipv6-wg/

**Comment by zajdee** [comment]
Now that #29472 is merged, CLAT is the last part missing in supporting of the IPv6-mostly networks.

**Comment by Leseratte10** [comment]
Just updated my Arch Linux box to the latest systemd with #29472 (RFC8925 support) included, I noticed a small issue related to the missing 464XLAT support. 

I'm not sure if I'm misunderstanding something (which may well be the case) or if my issue is caused by an Arch Linux oddity, in that case I apologize for wasting your time.

RFC 8925, which defines the new "IPv6 only" parameter, describes a "host [that is] capable of operating in IPv6 only mode" (which is signalled by this new parameter) to be a host that either only has IPv6-only-capable applications, or a host that's running a 464XLAT. 

Quote from RFC:
> **IPv6-only-capable host**: A host that does not require an IPv4 address and can operate on IPv6-only networks. More precisely, IPv6-only capability is specific to a given interface of the host: if some applications on a host require IPv4 and the 464XLAT CLAT (customer-side translator) [[RFC6877](https://datatracker.ietf.org/doc/html/rfc6877)] is only enabled on one interface, the host is IPv6-only capable if connected to a NAT64 network via that interface.

That definition makes it pretty clear that a host that has applications that require IPv4 (which isn't that uncommon even in 2023, sadly) but without a working 464XLAT is not to be considered an "IPv6-only capable host".

The IPv6-only parameter is intended for a network where there's both clients like Android and iOS (that auto-deploy their own 464XLAT when on an IPv6-only network) and clients like Windows, OSX, Linux, whatever that don't auto-deploy a 464XLAT and thus need native IPv4 connectivity, so that only clients in that 2nd group will receive an IPv4 address over DHCP. 

My network's DHCPv4 server has been running with the "IPv6-only" flag for a long time - to ensure that smartphones and tablets don't waste IPv4s they don't need. This means, that my Arch Linux box with a pretty standard network config (IPv4 enabled, IPv6 enabled, done) did no longer receive an IPv4 address using DHCP after this systemd update was installed. However, there's no code in systemd (yet) to automatically trigger and setup a 464XLAT in this case (as evidenced by this particular issue #23674 still being open). 

This means that while my IPv6-capable applications on the box were able to access the internet as usual (either using native IPv6 or using DNS64/NAT64), any applications that require IPv4 stopped working after this update until I manually set `IPv6OnlyMode=no` in the network config. 

Is there anything I could be missing or anything I did wrong? To me, this looks like systemd isn't implementing RFC8925 correctly since there's no automatic 464XLAT as required by the RFC, and this issue about the missing 464XLAT should have been solved / implemented before the IPv6-only flag was enabled / handled by default.

In my opinion (but this could be wrong, I'm not an expert), the default value of `IPv6OnlyMode` should be set to False, and only be changed to true once systemd has support for 464XLAT.

**Comment by silkeh** [comment]
@Leseratte10: 

> That definition makes it pretty clear that a host that has applications that require IPv4 (which isn't that uncommon even in 2023, sadly) but without a working and enabled 464XLAT with a NAT64 prefix either manually configured or automatically learned through RFC 7050 (DNS), RFC 7225 (PCP) or RFC 8781 (RA/PREF64) is not to be considered an "IPv6-only capable host".

This honestly seems like an issue for your distribution. They manage the packages in the repository and can change the systemd defaults based on the software in the repository. The distro is also in charge of software integration, which seems to be your issue. How would the systemd project know that you have or can install applications that rely on IPv4 connectivity?

> clients like Windows, OSX, Linux, whatever that don't auto-deploy a 464XLAT and thus need native IPv4 connectivity

Most of my deployments definitely don't need native IPv4 connectivity. Applications that actually require native IPv4 connectivity (and don't work with DNS64/NAT64) are fairly rare in my experience. Additionally, I don't know of any case where the application actually *needs* native IPv4 connectivity, instead of it being a bug in the software. This is especially the case for servers, where it is even relatively easy to run systems without connectivity to the IPv4 internet entirely (so no DNS64/NAT64).

**Comment by Leseratte10** [comment]
Okay, that does kinda make sense. On Desktop Linux it's more likely for software to be installed outside of the repositories, including IPv4-only software (Discord or something) while on server Linux it's more likely that a server would be fine with IPv6-only operation. I'll keep a look at a couple distros about what they'll do with that parameter once they update to 255. 

Nevertheless, I'm hoping that 464XLAT will get a higher priority for systemd and will be implemented soon-ish, if possible. Once that's the case this whole issue will solve itself, no matter which distributions enable or disable that default setting.

**Comment by patrakov** [comment]
No out-of-repositories software is required to trigger the failure. Firefox is enough.

I know of one case where native IPv4 connectivity is required: SIP and WebRTC with peers that don't support IPv6. For example, in Firefox, on an IPv6-only connection with DNS64, visit https://janus.conf.meetecho.com/echotest.html - the echo test fails. Start clatd (a Linux 464XLAT implementation), and it succeeds.

I think that at this stage video conferencing over the Internet is important enough that we should not force a configuration that breaks it. In other words, @Leseratte10 's network configuration is valid and should result in working IPv4 and IPv6 connectivity, including with IPv4 literals.

---

The error in the browser console is: `WebRTC: ICE failed, add a TURN server and see about:webrtc for more details`

One of the complicating factors is that I have a Tailscale VPN to a corporate network and some IPv4-only bridges for virtual machines, so I do have an IPv4 address, but no default IPv4 route unless I start clatd.

![image](https://github.com/systemd/systemd/assets/2943163/81c78e93-04e9-4e02-8c32-7dc09e95bf0c)

**Comment by Leseratte10** [comment]
That's what I mean, yeah. 

Given the huge amount of IPv6-incompatible software out there, it doesn't seem like a good idea to open bug reports for all of them in my distribution. As much as we'd all like to have IPv6 support, that's for the software devs to add, not the distribution maintainers. 

Yes, Firefox isn't usually used on servers and thus for servers it might make sense to honor the IPv6-only flag from DHCPv4 even when there is no supported 464XLAT, even though RFC 8925 says there should be one. But this is going to break every default installation on any Linux client. 

Are you saying that every single Linux distribution that's not only intended for just servers but either for both servers and clients or just for clients is supposed to alter systemd's default config to explicitly set `IPv6OnlyMode=false`? If so, this should have been mentioned in the Changelog's "Incompatible changes" section, imo, because it will break existing, perfectly sensible configurations.

Looking at the changelog right now, this change doesn't even seem to be mentioned in the changelog on the Github release for version 255 at all? How are distributions supposed to know / figure out they need to change this default config?

Wouldn't it be more sensible to change systemd's default to `IPv6OnlyMode=false` as a short-term solution and then set it back to true when a working 464XLAT is implemented eventually?

**Comment by SoniEx2** [comment]
what about implementing the CLAT (without the use of a daemon) in glibc instead?

**Comment by Leseratte10** [comment]
In glibc? I'm not that well versed in linux networking so maybe I'm talking crap here, but I don't think that that's a good idea. Then you'd need a whole bunch of NAT64 prefix discovery code in glibc (RFC 7050, 7225, 8781 and probably more), or a way to do prefix discovery in systemd and somehow let the glibc know about the prefix.

**Comment by SoniEx2** [comment]
it would be a great idea. don't bother with DNS64-based prefix discovery tho, DNS64 is a defective standard, since it breaks DNSSEC, but you don't need to worry about that if you have a CLAT.

**Comment by patrakov** [comment]
> what about implementing the CLAT (without the use of a daemon) in glibc instead?

This would not help, due to containers, Flatpaks, and whatnot.

**Comment by SoniEx2** [comment]
what about them? do those use NAT for networking?

**Comment by patrakov** [comment]
It depends. Docker containers can use either NAT or stay in the host network namespace. Flatpak uses the host network namespace.

**Comment by SoniEx2** [comment]
oh, right, you'd want to make sure the flatpak/container glibc is up-to-date with the latest CLAT changes.

(DNS64 is no longer appropriate for discovering the prefix. use the RAs. PREF64 or whatever.)

**Comment by Leseratte10** [comment]
> you'd want to make sure the flatpak/container glibc is up-to-date with the latest CLAT changes

Which can take a looong time, given that quite a few people are using containers as like "I have this ancient application that needs an ancient OS that I want to run on a modern OS" and stuff it into a container with an old base image.

**Comment by pmarks-net** [comment]
I agree with @Leseratte10. `IPv6OnlyMode=yes` violates RFC 8925 and should be changed.

Android and iOS were the pioneers in this space.  Android has 464XLAT and iOS tests applications for NAT64 compatibility as part of a manual review process.

Generic Linux doesn't have a curated app store, so the only RFC-compliant solution is to set `IPv6OnlyMode=no` until 464XLAT is available.  People have enough reasons to blame IPv6 for breaking their stuff. Please don't give them more.

**Comment by momoka0122y** [comment]
> Could you elaborate more? I am not familiar with 464XLAT. Could you suggest a setting format to support the feature? Also, could you provide a way to support that feature with another tools, e.g. `ip` command?

I don't know the systemd development method so I'm curious.
When systemd implements 464XLAT can it use a translation software (tayga or jool_siit) or would it need to build the CLAT from scrach?

I've been running a CLAT on my Ubuntu with jool_siit and it's working pretty nice so far.
Thanks to @oskar456 https://gist.github.com/oskar456/d898bf2e11b642757800a5ccdc2415aa
```
ip netns add clat-netns

ip link add name HOST_TO_CLAT4 type veth peer CLAT_IPv4_LINK
ip link set dev CLAT_IPv4_LINK netns clat-netns
ip link set up dev HOST_TO_CLAT4

ip link add name CLAT_IPv6_LINK link <interface> type ipvlan
ip link set dev CLAT_IPv6_LINK netns clat-netns

ip -6 neigh add proxy 2001:db8::c1a7 dev <interface>
sysctl net.ipv6.conf.<interface>.proxy_ndp=1

ip -4 address add 192.0.0.1/30 dev HOST_TO_CLAT4
ip -4 route add default via 192.0.0.2 dev HOST_TO_CLAT4

ip netns exec clat-netns ip link set up dev CLAT_IPv4_LINK
ip netns exec clat-netns ip link set up dev CLAT_IPv6_LINK
ip netns exec clat-netns ip -4 address add 192.0.0.2/30 dev CLAT_IPv4_LINK
ip netns exec clat-netns ip -6 address add 2001:db8::c1a7 dev CLAT_IPv6_LINK
ip netns exec clat-netns ip -6 route add default dev CLAT_IPv6_LINK
ip netns exec clat-netns jool_siit instance add --netfilter --pool6 64:ff9b::/96
ip netns exec clat-netns jool_siit eamt add 2001:db8::c1a7 192.0.0.1
```

Maybe not the most optimal way to do this but this works for me.
But this is configuring the NAT64 prefix and it would be nice if systemd could get that information from the Router Advertisements.

So I don't know what everyone else wants but a configuration I would like might look something like
```
[464XLAT]
# Enables the 464XLAT feature when a PREF64 is detected in Router Advertisements (RA).
EnableOnPREF64=yes

# Specifies the CLAT IPv4 address to use.
CLATAddress=192.0.0.1/30

# Forces the 464XLAT feature to be enabled, even if no PREF64 is received from RA.
ForceEnable=yes

# Specifies the NAT64 prefix to be used when ForceEnable is set to 'yes'.
ForceEnableNAT64Prefix=64:ff9b::/96
```

**Comment by Leseratte10** [comment]
I don't know, and I don't think I'm qualified enough to answer that question. In the end, for most users, it won't matter. Either systemd implements its own CLAT functionality, or it just includes the code from Tayga or Jool or [clatd](https://github.com/toreanderson/clatd). 

It just should do something, because the current version breaks RFC and breaks people's networks. Whether that "something" is "adding a CLAT" or "disabling `IPv6OnlyMode`" is up to the systemd developers.

**Comment by oskar456** [comment]
Hey, I couldn't agree more with @Leseratte10 that `IPv6OnlyMode` should **default to disabled,** until CLAT is implemented and enabled in systemd. The whole point of RFC 8925 is to have just one network, which is used as IPv6-only for devices featuring CLAT (so that the IPv4 connectivity is established via CLAT) and as IPv4/IPv6 dual-stack for devices without CLAT. If systemd by default opt out from native IPv4 while not providing CLAT feature, this leads to breakage of IPv4-only dependent applications which is exactly what the authors of RFC 8925 wanted to avoid.

If such breakage was acceptable, then the network could have been run as IPv6-only without any IPv4 support. The whole point of running IPv4 DHCP with option 108 support is to provide IPv4 connectivity for devices without CLAT. Such devices should not opt out from IPv4.

It would be nice to see a support for configuring CLAT but since the SIIT translation between IPv4 and IPv6 is not part of Linux, this brings a dependency on a third party code like `jool` or `TAYGA`. Therefore, even if systemd supports setting up CLAT (which I would really like to see), that feature will probably be off by default and only enabled if particular 3rd party software is installed.

**Comment by Leseratte10** [comment]
@yuwata Sorry for the ping, but since you were the one who first triaged this report as a feature request and none of the systemd developers seem to have commented on this issue since I reported it two weeks ago - can this report be reclassified as a bug instead of a feature request to hopefully give it a bit more priority?

A clear violation of an RFC should, in my opinion, no longer be classified as a "request for enhancement". Instead it deserves the label "bug" and possibly even "regression" since the bug was just introduced in the last release and it breaks on networks where it worked before without any issue.

**Comment by silkeh** [comment]
@Leseratte10: Maybe it would be more useful to create a new issue for the specific problem instead of hijacking this one?

This issue is about 464XLAT support an (imho) does not really have much to do with the default value of `IPv6OnlyMode`, except for that the default could be switched after a CLAT has been implemented.

**Comment by Leseratte10** [comment]
Okay, will do.

**Comment by certuna** [comment]
> (DNS64 is no longer appropriate for discovering the prefix. use the RAs. PREF64 or whatever.)

[RFC 7050](https://datatracker.ietf.org/doc/rfc7050/) is still valid, discovering the prefix through DNS (`ipv4only.arpa`) is allowed.

I do agree it's not ideal, since IPv4 connectivity will break as soon as a client machine (or application) uses a custom DNS server, but it's not deprecated.

**Comment by SoniEx2** [comment]
yes and more importantly we're personally trying to get ppl to stop breaking DNSSEC. :)

**Comment by certuna** [comment]
RFC7050 doesn't break DNSSEC, it's only for discovering the prefix. What you're thinking of is DNS64 where the synthesized AAAA records are generated by the DNS server so they can't be validated by the client.

A DNS server with `ipv4only.arpa` record does not need to synthesize any AAAA records, this can be done by the client _after_ validating DNSSEC.

**Comment by ogelpre** [comment]
Microsoft plans to bring [general CLAT support for Windows](https://techcommunity.microsoft.com/t5/networking-blog/windows-11-plans-to-expand-clat-support/ba-p/4078173). Linux is falling behind.

Besides the already mentioned CLAT implementations OpenWRT has it's own CLAT implementation written in C. The package is called [464xlat](https://github.com/openwrt/openwrt/tree/openwrt-22.03/package/network/ipv6/464xlat).

**Comment by ogelpre** [comment]
Besides the already mentioned CLAT implementations there is a [kernel module implementation](https://github.com/ayourtch/nat46) which is used by OpenWRT.

**Comment by pyfisch** [comment]
There is a new RFC draft titled [464XLAT Customer-side Translator (CLAT): Node Recommendations](https://datatracker.ietf.org/doc/draft-ietf-v6ops-claton/) that makes some recommendations how CLAT should be implemented on devices ("wireless network architecture") similar to the existing CLAT support on Apple devices and Android phones.

I believe the existing CLAT implementations for Linux focus on router use-cases  ("wireline network architecture") and will need some modifications before they can be used by systemd. Notably a CLAT running on a Linux laptop needs to dynamically enable and disable CLAT based on DHCPv4, Router Advertisements and changing network interfaces.

**Comment by Leseratte10** [comment]
That seems to be a sensible draft, and it clears up the question about when to start the CLAT process and whether to prioritize IPv6 over IPv4, and predictably, it's like "Get IPv6 and the CLAT up and running as fast as possible and if you do receive an IPv4 address over DHCP just stop the CLAT again", which means IPv6 is prioritized and set-up first. 

I hope that when this RFC becomes a proper standard it'll lead to more tools and systems (like systemd) to finally support IPv6-only networks properly.

**Comment by GoetzGoerisch** [comment]
This is probably the requirement after which a CLAT for LAN and WIFI is integrated into Windows 11 and ChromeOS going forward.

**Comment by toreanderson** [comment]
What first needs to happen here is that Someone™ needs to go through the trouble of getting an implementation of [SIIT](https://datatracker.ietf.org/doc/html/rfc7915)/[SIIT-EAM](https://datatracker.ietf.org/doc/html/rfc7757) into the upstream Linux kernel. Once a packet translation engine is available there, adding the necessary bits to systemd-networkd (and NetworkManager, ifupdown, etc. for that matter) should be low-hanging fruit.

**Comment by GoetzGoerisch** [comment]
NLnet is funding a project aiming for this: https://nlnet.nl/project/IPv6-monostack/

**Comment by pyfisch** [comment]
@GoetzGoerisch According to NLnet the project started in 2023-12, however the website is down and a cursory search of kernel.org doesn't find relevant patches. Is there anything publicly available yet? 

I think the SIIT can be implemented either in kernel or userspace.

**Comment by pemensik** [comment]
> What first needs to happen here is that Someone™ needs to go through the trouble of getting an implementation of [SIIT](https://datatracker.ietf.org/doc/html/rfc7915)/[SIIT-EAM](https://datatracker.ietf.org/doc/html/rfc7757) into the upstream Linux kernel. Once a packet translation engine is available there, adding the necessary bits to systemd-networkd (and NetworkManager, ifupdown, etc. for that matter) should be low-hanging fruit.

I do not think it needs to be done in Linux kernel neccesarily. After all, it is fallback only. If raw performance is desired, just implement native IPv6. If it is slower but generally working, I think userspace is okay for that.

I think main problem is missing integration with existing userspace implementations.

**Comment by toreanderson** [comment]
> I do not think it needs to be done in Linux kernel neccesarily. After all, it is fallback only. If raw performance is desired, just implement native IPv6. If it is slower but generally working, I think userspace is okay for that.
> 
> I think main problem is missing integration with existing userspace implementations.

From a technical standpoint, you are correct. That said, I highly doubt there is much appetite for adding a dependency from systemd-networkd (or NetworkManager or whatever) on a user-space implementation like [TAYGA](http://www.litech.org/tayga/) (which appears to be unmaintained). I am not aware of any other usable user-space implementations of SIIT/SIIT-EAM.

**Comment by SoniEx2** [comment]
why *depend* on TAYGA? fork it *into* systemd-networkd!

**Comment by patrakov** [comment]
There is also https://github.com/xdp-project/bpf-examples/blob/master/nat64-bpf/README.org

**Comment by SoniEx2** [comment]
there, fixed. hope that helps.

**Comment by pemensik** [comment]
Please, not a new systemd-only reimplementation again. We have already enough of such services. We need maintained and usable implementation. https://github.com/vitlabuda/tundra-nat64 might be another solution. There might be others. But bundling the service in systemd project will not bring people working on it magically. It would mean they will not work on something else already present in systemd project in the best case.

It needs runtime reconfiguration API, ideally [libeconf](https://github.com/openSUSE/libeconf) for configuration. But it does not have to be part of systemd. It needs to be integrated with systemd, but that is not the same thing. I would prefer NetworkManager and systemd-networkd would share the same maintained implementation, just with different frontends used. Instead of fragmenting efforts please merge them.

**Comment by GoetzGoerisch** [comment]
@pemensik at RIPE89, it was told that a solution, if prioritised by RH customers is in works? Do you have a pointer to show to customers?

**Comment by SoniEx2** [comment]
how hard can it be to maintain a packet translator... it's mostly set and forget, that's why tayga still works today and doesn't need maintenance. on the other hand, the systemd integration needs active maintenance, as the requirements between systemd and the packet translator can change.

this is why it makes so much sense to just put it in systemd. and possibly in networkmanager. (does networkmanager bring its own dhcp yet?) if you want something standalone, just use something unmaintained, like tayga.

**Comment by pemensik** [comment]
@GoetzGoerisch I do not work in Network Manager or systemd team and I am not aware of any planned changes in this area. Sorry.

What is missing is a good way to setup the interface, bring clat up and down via runtime reconfiguration, ideally DBus or varlink driven. Every solution so far needed manual configuration, at least what I read about that. That is not usable with PREF64 option in such case.

**Comment by SoniEx2** [comment]
@pemensik wait, do you even need to? can't you just pop routes...?

**Comment by SoniEx2** [comment]
alternatively: put the CLAT in the libc, who needs syscall compatibility anyway.

**Comment by GoetzGoerisch** [comment]
> @GoetzGoerisch According to NLnet the project started in 2023-12, however the website is down and a cursory search of kernel.org doesn't find relevant patches. Is there anything publicly available yet?

I was in contact with one of the developers earlier and asked him to add a comment here.

**Comment by Sporesirius** [comment]
@GoetzGoerisch any news on that?

**Comment by DanielG** [comment]
> I was in contact with one of the developers

That would be me 👋 

A kernel implementation is in the works, but everyone please be patient. The project will be widely announced when we're ready to post patches. Watch netdev, RIPE v6-wg, IETF v6ops or debian-ipv6 lists for updates.

--Daniel

**Comment by SoniEx2** [comment]
relatedly: it looks like there's work underway to put a CLAT in NetworkManager. as it should be.

**Comment by artizirk** [comment]
NetworkManager CLAT merge request 

https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/2107

**Comment by tokred** [comment]
@DanielG I see some progress here (https://codeberg.org/IPv6-Monostack/joolif-net-next), any chance that's the project you are referring to?

**Comment by AsGreyWolf** [comment]
Looks like the 464xlat impl requires some kind of prefix storage (e.g. for deligated prefixes, for pref64, they look first-hop-bound), 
Rule 5.5 of source address selection (rfc6724) as well as rfc8028 requires prefix storage as well
I wish we could have a common iface+first-hop bound prefix storage in linux's netlink structures

**Comment by artizirk** [comment]
NetworkManager CLAT has been merged https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/2107

**Comment by artizirk** [comment]
> Introducing ipxlat: a stateless IPv4/IPv6 translation device

https://lore.kernel.org/netdev/20260319151230.655687-1-ralf@mandelbit.com/
